pkg/controller/podlocality/controller.go

```diff
@@ -0,0 +1,181 @@
+			return evt.ObjectOld != nil && evt.ObjectNew != nil &&
+				evt.ObjectOld.(*v1.Pod).Spec.NodeName == "" && evt.ObjectNew.(*v1.Pod).Spec.NodeName != ""
+		},
+		GenericFunc: func(evt event.GenericEvent) bool { return false },
+	})
+	if err != nil {
+		return err
+	}
+
+	err = mgr.GetFieldIndexer().IndexField(&v1.Pod{}, "spec.nodeName", func(obj apimachineryruntime.Object) []string {
+		pod := obj.(*v1.Pod)
+		return []string{pod.Spec.NodeName}
+	})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &v1.Node{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(a handler.MapObject) []reconcile.Request {
+			list := &v1.PodList{}
+			err := mgr.GetClient().List(context.TODO(), client.MatchingField("spec.nodeName", a.Meta.GetName()), list)
+			if err != nil {
+				log.Error(err, "Could not list pods")
+			}
+
+			var requests []reconcile.Request
+			for _, pod := range list.Items {
+				requests = append(requests, reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      pod.Name,
+						Namespace: pod.Namespace,
+					},
+				})
+			}
+			return requests
+		}),
+	}, predicate.Funcs{
+		CreateFunc: func(evt event.CreateEvent) bool { return evt.Meta != nil },
+		DeleteFunc: func(evt event.DeleteEvent) bool { return false },
+		UpdateFunc: func(evt event.UpdateEvent) bool {
+			return evt.MetaOld != nil && evt.MetaNew != nil &&
+				(evt.MetaOld.GetLabels()[NodeRegionLabel] != evt.MetaNew.GetLabels()[NodeRegionLabel] ||
+					evt.MetaOld.GetLabels()[NodeZoneLabel] != evt.MetaNew.GetLabels()[NodeZoneLabel])
+		},
+		GenericFunc: func(evt event.GenericEvent) bool { return false },
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &PodLocalityReconciler{}
+
+// PodLocalityReconciler copies the node's region/zone labels to the pod after it's scheduled to a node
+type PodLocalityReconciler struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	common.ResourceManager
+	scheme *runtime.Scheme
+}
+
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *PodLocalityReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Processing Pod")
+
+	// Fetch the Pod
+	pod := &v1.Pod{}
+	err := r.Client.Get(context.TODO(), request.NamespacedName, pod)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object
+		return reconcile.Result{}, err
+	}
+
+	if _, hasSidecar := pod.Annotations[IstioSidecarStatusAnnotation]; !hasSidecar {
```

Comment:
Should this be moved into the filter predicate funcs?

Reply:
Would still need to be here because I'm enqueuing pods whenever a node's zone/region labels change. But I could add the same check to the filter predicates.

Reply:
You could just not queue those pods, placing the annotation check in the node watcher too.

Reply:
Ah, of course. I forgot I was already listing pods there, so I was able to easily add that there, yes.